concordium-smart-contract-testing/src/lib.rs

```diff
@@ -0,0 +1,2745 @@
+use anyhow::bail;
+use concordium_base::{
+    base::{Energy, ExchangeRate},
+    transactions::{self, cost},
+};
+use concordium_contracts_common::*;
+use sha2::{Digest, Sha256};
+use std::{
+    collections::{
+        btree_map::Entry::{Occupied, Vacant},
+        BTreeMap,
+    },
+    path::Path,
+};
+use thiserror::Error;
+use wasm_chain_integration::{
+    v0,
+    v1::{
+        self,
+        trie::{MutableState, SizeCollector},
+        ConcordiumAllowedImports, InvokeResponse, ReturnValue,
+    },
+    ExecResult, InterpreterEnergy,
+};
+use wasm_transform::artifact;
+
+/// A V1 artifact, with concrete types for the generic parameters.
+type ArtifactV1 = artifact::Artifact<v1::ProcessedImports, artifact::CompiledFunction>;
+
+// Energy constants from Cost.hs in concordium-base.
+const QUERY_ACCOUNT_BALANCE_COST: u64 = 200;
+const QUERY_CONTRACT_BALANCE_COST: u64 = 200;
+const QUERY_EXCHANGE_RATE_COST: u64 = 100;
+const INITIALIZE_CONTRACT_INSTANCE_BASE_COST: u64 = 300;
+const INITIALIZE_CONTRACT_INSTANCE_CREATE_COST: u64 = 200;
+
+pub struct Chain {
+    /// The slot time viewable inside the smart contracts.
+    /// Defaults to `0`.
+    slot_time:           SlotTime,
+    /// MicroCCD per Euro ratio.
+    micro_ccd_per_euro:  ExchangeRate,
+    /// Euro per Energy ratio.
+    euro_per_energy:     ExchangeRate,
+    /// Accounts and info about them.
+    accounts:            BTreeMap<AccountAddress, AccountInfo>,
+    /// Smart contract modules.
+    modules:             BTreeMap<ModuleReference, ArtifactV1>,
+    /// Smart contract instances.
+    contracts:           BTreeMap<ContractAddress, ContractInstance>,
+    /// Next contract index to use when creating a new instance.
+    next_contract_index: u64,
+}
+
+#[derive(Clone)]
+pub struct ContractInstance {
+    module_reference: ModuleReference,
+    contract_name:    OwnedContractName,
+    state:            v1::trie::PersistentState,
+    owner:            AccountAddress,
+    self_balance:     Amount,
+}
+
+impl Chain {
+    pub fn new_with_time_and_rates(
+        slot_time: SlotTime,
+        micro_ccd_per_euro: ExchangeRate,
+        euro_per_energy: ExchangeRate,
+    ) -> Self {
+        Self {
+            slot_time,
+            micro_ccd_per_euro,
+            euro_per_energy,
+            accounts: BTreeMap::new(),
+            modules: BTreeMap::new(),
+            contracts: BTreeMap::new(),
+            next_contract_index: 0,
+        }
+    }
+
+    pub fn new_with_time(slot_time: SlotTime) -> Self {
+        Self {
+            slot_time,
+            ..Self::new()
+        }
+    }
+
+    /// Create a new [`Self`] where
+    ///  - `slot_time` defaults to `0`,
+    ///  - `micro_ccd_per_euro` defaults to `16036807715944130560 /
+    ///    108919627567`,
```

Comment:
It was the current chain value when writing this code ;) I subsequently simplified the ratio to `147235241 / 1` because I got an overflow when converting NRG to CCD. There is a TODO regarding that.